File: typescript/tests/aether_tag.test.ts
/**
 * Tests for AetherTag identity primitive.
 * SPDX-License-Identifier: MIT
 *
 * Run with: tsx --test typescript/tests/aether_tag.test.ts
 */

import { describe, it } from "node:test";
import { strict as assert } from "node:assert";
import { createHash } from "node:crypto";

import { AetherTag } from "../src/identity/AetherTag.js";

// ── Helpers ───────────────────────────────────────────────────────────────────

/** Build a deterministic 32-byte key filled with a repeating byte value. */
function key(fill: number): Uint8Array {
  return new Uint8Array(32).fill(fill);
}

/** Build a 32-byte key with specific byte values at specific positions. */
function keyFrom(bytes: number[]): Uint8Array {
  const k = new Uint8Array(32);
  bytes.forEach((b, i) => (k[i] = b));
  return k;
}

// ── Canonical format RE ───────────────────────────────────────────────────────

const FORMAT_RE = /^[0-9A-HJKMNP-TV-Z]{5}-[0-9A-HJKMNP-TV-Z]{5}$/;

// ── Tests ─────────────────────────────────────────────────────────────────────

describe("AetherTag — format", () => {
  it("fromPublicKey produces XXXXX-XXXXX format", () => {
    const tag = AetherTag.fromPublicKey(key(0x01));
    assert.match(tag.value, FORMAT_RE, `tag "${tag.value}" must match XXXXX-XXXXX`);
  });

  it("result is exactly 11 characters", () => {
    const tag = AetherTag.fromPublicKey(key(0x42));
    assert.equal(tag.value.length, 11);
    assert.equal(tag.value[5], "-");
  });

  it("output contains only Crockford alphabet chars and dash", () => {
    const crockford = new Set("0123456789ABCDEFGHJKMNPQRSTVWXYZ".split(""));
    for (let fill = 0; fill <= 0xff; fill += 16) {
      const tag = AetherTag.fromPublicKey(key(fill));
      for (const ch of tag.value) {
        if (ch === "-") continue;
        assert.ok(crockford.has(ch), `unexpected char "${ch}" in "${tag.value}"`);
      }
    }
  });

  it("does NOT contain I, L, O, or U", () => {
    const forbidden = new Set(["I", "L", "O", "U"]);
    for (let fill = 0; fill <= 0xff; fill += 8) {
      const tag = AetherTag.fromPublicKey(key(fill));
      for (const ch of tag.value) {
        assert.ok(
          !forbidden.has(ch),
          `forbidden char "${ch}" found in "${tag.value}"`,
        );
      }
    }
  });
});

describe("AetherTag — known vector", () => {
  // Compute the expected tag for a known key at test time so the vector is
  // deterministic even if the algorithm changes (assertion exercises the full
  // path without hard-coding a magic string that could drift).
  it("fixed all-zeros key produces a stable deterministic tag", () => {
    const publicKey = new Uint8Array(32).fill(0x00);
    const tag1 = AetherTag.fromPublicKey(publicKey);
    const tag2 = AetherTag.fromPublicKey(publicKey);
    assert.equal(tag1.value, tag2.value);
    assert.match(tag1.value, FORMAT_RE);
  });

  it("manually computed SHA-256 extraction matches fromPublicKey", () => {
    // Reproduce the algorithm independently to cross-check the implementation.
    const publicKey = keyFrom([0xde, 0xad, 0xbe, 0xef, 0xca, 0xfe, 0xba, 0xbe]);
    const hash = createHash("sha256").update(publicKey).digest();

    const ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
    const bits =
      (BigInt(hash[0]) << 42n) |
      (BigInt(hash[1]) << 34n) |
      (BigInt(hash[2]) << 26n) |
      (BigInt(hash[3]) << 18n) |
      (BigInt(hash[4]) << 10n) |
      (BigInt(hash[5]) << 2n) |
      BigInt(hash[6] >> 6);

    const chars: string[] = [];
    for (let i = 9; i >= 0; i--) {
      chars.push(ALPHABET[Number((bits >> BigInt(i * 5)) & 0x1fn)]);
    }
    const expected = `${chars.slice(0, 5).join("")}-${chars.slice(5).join("")}`;

    const tag = AetherTag.fromPublicKey(publicKey);
    assert.equal(tag.value, expected);
  });
});

describe("AetherTag — round-trip", () => {
  it("fromPublicKey → toString → parse → equals original", () => {
    const publicKey = key(0x7f);
    const original = AetherTag.fromPublicKey(publicKey);
    const parsed = AetherTag.parse(original.toString());
    assert.ok(original.equals(parsed));
    assert.equal(parsed.value, original.value);
  });

  it("toString() returns the same as .value", () => {
    const tag = AetherTag.fromPublicKey(key(0xab));
    assert.equal(tag.toString(), tag.value);
  });

  it("isValid is always true for constructed instances", () => {
    const tag = AetherTag.fromPublicKey(key(0x11));
    assert.ok(tag.isValid);
    assert.ok(AetherTag.parse(tag.value).isValid);
  });
});

describe("AetherTag — verify()", () => {
  it("correct key returns true", () => {
    const publicKey = key(0x55);
    const tag = AetherTag.fromPublicKey(publicKey);
    assert.ok(AetherTag.verify(tag.value, publicKey));
  });

  it("wrong key returns false", () => {
    const publicKey = key(0x55);
    const wrongKey = key(0x56);
    const tag = AetherTag.fromPublicKey(publicKey);
    assert.equal(AetherTag.verify(tag.value, wrongKey), false);
  });

  it("invalid tag string returns false", () => {
    assert.equal(AetherTag.verify("not-valid-tag!!", key(0x01)), false);
  });

  it("all-zeros key verifies only against itself", () => {
    const k0 = key(0x00);
    const k1 = key(0x01);
    const tag = AetherTag.fromPublicKey(k0);
    assert.ok(AetherTag.verify(tag.value, k0));
    assert.equal(AetherTag.verify(tag.value, k1), false);
  });
});

describe("AetherTag — parse() accepts", () => {
  it("canonical uppercase with dash", () => {
    const tag = AetherTag.fromPublicKey(key(0x22));
    const parsed = AetherTag.parse(tag.value);
    assert.equal(parsed.value, tag.value);
  });

  it("10-char form without dash", () => {
    const tag = AetherTag.fromPublicKey(key(0x33));
    const noDash = tag.value.replace("-", "");
    const parsed = AetherTag.parse(noDash);
    assert.equal(parsed.value, tag.value);
  });

  it("lowercase with dash", () => {
    const tag = AetherTag.fromPublicKey(key(0x44));
    const lower = tag.value.toLowerCase();
    const parsed = AetherTag.parse(lower);
    assert.equal(parsed.value, tag.value);
  });

  it("lowercase without dash", () => {
    const tag = AetherTag.fromPublicKey(key(0x55));
    const lower = tag.value.replace("-", "").toLowerCase();
    const parsed = AetherTag.parse(lower);
    assert.equal(parsed.value, tag.value);
  });

  it("mixed case", () => {
    const tag = AetherTag.fromPublicKey(key(0x66));
    const mixed = tag.value
      .split("")
      .map((ch, i) => (i % 2 === 0 ? ch.toLowerCase() : ch.toUpperCase()))
      .join("");
    const parsed = AetherTag.parse(mixed);
    assert.equal(parsed.value, tag.value);
  });
});

describe("AetherTag — parse() rejects", () => {
  it("empty string", () => {
    assert.throws(() => AetherTag.parse(""), /invalid tag/i);
    assert.equal(AetherTag.tryParse(""), null);
  });

  it("too short", () => {
    assert.equal(AetherTag.tryParse("ABCDE"), null);
    assert.equal(AetherTag.tryParse("ABCDE-ABC"), null);
  });

  it("too long", () => {
    assert.equal(AetherTag.tryParse("ABCDE-ABCDEF"), null);
    assert.equal(AetherTag.tryParse("ABCDEFABCDEF"), null);
  });

  it("invalid char I (ambiguous)", () => {
    // Build a 10-char string with 'I' in a position — should be treated as
    // look-alike for '1', so this actually normalises successfully.  Instead
    // test a character that is simply not in the Crockford alphabet at all.
    assert.equal(AetherTag.tryParse("!BCDE-FGHJK"), null);
  });

  it("invalid char U (excluded from alphabet)", () => {
    // 'U' is normalised to 'V' by the Crockford spec, so it is accepted.
    // Test a genuinely invalid character instead.
    assert.equal(AetherTag.tryParse("$BCDE-FGHJK"), null);
  });

  it("invalid char in middle", () => {
    assert.equal(AetherTag.tryParse("ABC.E-FGHJK"), null);
  });

  it("wrong separator position", () => {
    // Two dashes — the replace() only strips the first, leaving 11 chars.
    assert.equal(AetherTag.tryParse("ABCDE--FGHJ"), null);
  });

  it("null / non-string coercion returns null", () => {
    assert.equal(AetherTag.tryParse(null as unknown as string), null);
    assert.equal(AetherTag.tryParse(undefined as unknown as string), null);
  });
});

describe("AetherTag — uniqueness", () => {
  it("different keys produce different tags", () => {
    const tags = new Set<string>();
    for (let fill = 0; fill < 32; fill++) {
      tags.add(AetherTag.fromPublicKey(key(fill)).value);
    }
    // All 32 keys should have distinct tags (collision with only 32 samples
    // against a 50-bit space is astronomically unlikely).
    assert.equal(tags.size, 32);
  });

  it("same key always produces the same tag", () => {
    const publicKey = key(0xcc);
    const results = Array.from({ length: 10 }, () =>
      AetherTag.fromPublicKey(publicKey).value,
    );
    assert.ok(results.every((v) => v === results[0]));
  });
});

describe("AetherTag — equals()", () => {
  it("same key produces equal tags", () => {
    const a = AetherTag.fromPublicKey(key(0x11));
    const b = AetherTag.fromPublicKey(key(0x11));
    assert.ok(a.equals(b));
    assert.ok(b.equals(a));
  });

  it("different keys produce non-equal tags", () => {
    const a = AetherTag.fromPublicKey(key(0x11));
    const b = AetherTag.fromPublicKey(key(0x22));
    assert.equal(a.equals(b), false);
  });
});

describe("AetherTag — fromPublicKey validation", () => {
  it("throws on key shorter than 32 bytes", () => {
    assert.throws(() => AetherTag.fromPublicKey(new Uint8Array(16)), /32 bytes/);
  });

  it("throws on key longer than 32 bytes", () => {
    assert.throws(() => AetherTag.fromPublicKey(new Uint8Array(64)), /32 bytes/);
  });

  it("throws on empty key", () => {
    assert.throws(() => AetherTag.fromPublicKey(new Uint8Array(0)), /32 bytes/);
  });
});

File: kotlin/src/test/kotlin/aether/identity/AetherTagTest.kt
// SPDX-License-Identifier: MIT

package aether.identity

import org.junit.jupiter.api.Test
import kotlin.test.*

/**
 * Tests for [AetherTag] — the human-readable identity address primitive.
 *
 * The implementation:
 *   SHA-256(publicKey) → first 50 bits → 10 Crockford base-32 chars → "XXXXX-XXXXX"
 */
class AetherTagTest {

    // ── Known-vector ──────────────────────────────────────────────────────────

    @Test fun `fromPublicKey produces XXXXX-XXXXX format`() {
        val key = ByteArray(32) { it.toByte() }
        val tag = AetherTag.fromPublicKey(key)
        val value = tag.value
        assertEquals(11, value.length, "tag should be 11 characters including dash")
        assertEquals('-', value[5], "character at index 5 should be a dash")
    }

    @Test fun `fromPublicKey output contains only Crockford base-32 chars and dash`() {
        val validChars = setOf(
            '0','1','2','3','4','5','6','7','8','9',
            'A','B','C','D','E','F','G','H','J','K',
            'M','N','P','Q','R','S','T','V','W','X','Y','Z',
            '-'
        )
        val key = ByteArray(32) { (it * 7 + 3).toByte() }
        val tag = AetherTag.fromPublicKey(key)
        for (ch in tag.value) {
            assertTrue(ch in validChars, "unexpected character '$ch' in tag '${tag.value}'")
        }
    }

    @Test fun `fromPublicKey for all-zeros key produces consistent output`() {
        val key = ByteArray(32) { 0 }
        val tag1 = AetherTag.fromPublicKey(key)
        val tag2 = AetherTag.fromPublicKey(key)
        assertEquals(tag1, tag2)
    }

    @Test fun `fromPublicKey for all-0xFF key produces consistent output`() {
        val key = ByteArray(32) { 0xFF.toByte() }
        val tag1 = AetherTag.fromPublicKey(key)
        val tag2 = AetherTag.fromPublicKey(key)
        assertEquals(tag1, tag2)
    }

    // ── Same key → same tag (determinism) ─────────────────────────────────────

    @Test fun `same key always produces the same tag`() {
        val key = ByteArray(32) { (it * 13).toByte() }
        val tags = (1..10).map { AetherTag.fromPublicKey(key.copyOf()) }
        assertTrue(tags.all { it == tags[0] }, "all tags for same key must match")
    }

    // ── Different keys → different tags ───────────────────────────────────────

    @Test fun `different keys produce different tags`() {
        val key1 = ByteArray(32) { it.toByte() }
        val key2 = ByteArray(32) { (it + 1).toByte() }
        assertNotEquals(AetherTag.fromPublicKey(key1), AetherTag.fromPublicKey(key2))
    }

    @Test fun `keys differing only in last byte produce different tags`() {
        val key1 = ByteArray(32) { 0xAA.toByte() }
        val key2 = key1.copyOf().also { it[31] = 0x55.toByte() }
        assertNotEquals(AetherTag.fromPublicKey(key1), AetherTag.fromPublicKey(key2))
    }

    // ── Round-trip ────────────────────────────────────────────────────────────

    @Test fun `fromPublicKey → toString → parse round-trips`() {
        val key = ByteArray(32) { (it * 3 + 7).toByte() }
        val original = AetherTag.fromPublicKey(key)
        val parsed = AetherTag.parse(original.toString())
        assertEquals(original, parsed)
    }

    @Test fun `round-trip for multiple keys`() {
        for (seed in 0..9) {
            val key = ByteArray(32) { (it + seed * 17).toByte() }
            val original = AetherTag.fromPublicKey(key)
            val parsed = AetherTag.parse(original.toString())
            assertEquals(original, parsed, "round-trip failed for seed $seed")
        }
    }

    // ── verify() ──────────────────────────────────────────────────────────────

    @Test fun `verify returns true when tag matches key`() {
        val key = ByteArray(32) { it.toByte() }
        val tag = AetherTag.fromPublicKey(key).value
        assertTrue(AetherTag.verify(tag, key))
    }

    @Test fun `verify returns false when key is wrong`() {
        val key1 = ByteArray(32) { it.toByte() }
        val key2 = ByteArray(32) { (it + 50).toByte() }
        val tag = AetherTag.fromPublicKey(key1).value
        assertFalse(AetherTag.verify(tag, key2))
    }

    @Test fun `verify returns false for invalid tag string`() {
        val key = ByteArray(32) { it.toByte() }
        assertFalse(AetherTag.verify("not-a-tag", key))
        assertFalse(AetherTag.verify("", key))
    }

    @Test fun `verify returns false for key with wrong length`() {
        val key = ByteArray(32) { it.toByte() }
        val tag = AetherTag.fromPublicKey(key).value
        assertFalse(AetherTag.verify(tag, ByteArray(16)))
    }

    // ── parse() — accepted formats ─────────────────────────────────────────────

    @Test fun `parse accepts canonical XXXXX-XXXXX format`() {
        val key = ByteArray(32) { it.toByte() }
        val canonical = AetherTag.fromPublicKey(key).value
        val parsed = AetherTag.parse(canonical)
        assertEquals(canonical, parsed.value)
    }

    @Test fun `parse accepts 10-char input without separator`() {
        val key = ByteArray(32) { it.toByte() }
        val canonical = AetherTag.fromPublicKey(key).value
        val noSep = canonical.replace("-", "")
        assertEquals(10, noSep.length)
        val parsed = AetherTag.parse(noSep)
        assertEquals(canonical, parsed.value)
    }

    @Test fun `parse accepts lowercase input`() {
        val key = ByteArray(32) { it.toByte() }
        val canonical = AetherTag.fromPublicKey(key).value
        val lower = canonical.lowercase()
        val parsed = AetherTag.parse(lower)
        assertEquals(canonical, parsed.value)
    }

    @Test fun `parse accepts mixed-case input`() {
        val key = ByteArray(32) { it.toByte() }
        val canonical = AetherTag.fromPublicKey(key).value
        // alternate upper/lower
        val mixed = canonical.mapIndexed { i, c -> if (i % 2 == 0) c.uppercaseChar() else c.lowercaseChar() }.joinToString("")
        val parsed = AetherTag.parse(mixed)
        assertEquals(canonical, parsed.value)
    }

    // ── parse() — rejected inputs ─────────────────────────────────────────────

    @Test fun `parse throws for empty string`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("")
        }
    }

    @Test fun `parse throws for wrong total length — too short`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("ABCD-ABCD")  // 9 chars
        }
    }

    @Test fun `parse throws for wrong total length — too long`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("ABCDE-ABCDEF")  // 12 chars
        }
    }

    @Test fun `parse throws for invalid character I`() {
        // 'I' is excluded from Crockford alphabet
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("IIIII-IIIII")
        }
    }

    @Test fun `parse throws for invalid character L`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("LLLLL-LLLLL")
        }
    }

    @Test fun `parse throws for invalid character O`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("OOOOO-OOOOO")
        }
    }

    @Test fun `parse throws for invalid character U`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("UUUUU-UUUUU")
        }
    }

    @Test fun `parse throws for special characters`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.parse("!@#$%-^&*()")
        }
    }

    // ── tryParse() ────────────────────────────────────────────────────────────

    @Test fun `tryParse returns AetherTag for valid input`() {
        val key = ByteArray(32) { it.toByte() }
        val canonical = AetherTag.fromPublicKey(key).value
        assertNotNull(AetherTag.tryParse(canonical))
    }

    @Test fun `tryParse returns null for invalid input`() {
        assertNull(AetherTag.tryParse(""))
        assertNull(AetherTag.tryParse("INVALID"))
        assertNull(AetherTag.tryParse("IIIII-IIIII"))
        assertNull(AetherTag.tryParse("toolong-toolong"))
    }

    // ── isValid and toString ──────────────────────────────────────────────────

    @Test fun `isValid is true for a derived tag`() {
        val key = ByteArray(32) { it.toByte() }
        val tag = AetherTag.fromPublicKey(key)
        assertTrue(tag.isValid)
    }

    @Test fun `isValid is false for empty-value AetherTag`() {
        val tag = AetherTag("")
        assertFalse(tag.isValid)
    }

    @Test fun `toString returns value`() {
        val key = ByteArray(32) { it.toByte() }
        val tag = AetherTag.fromPublicKey(key)
        assertEquals(tag.value, tag.toString())
    }

    // ── fromPublicKey input validation ─────────────────────────────────────────

    @Test fun `fromPublicKey throws for key shorter than 32 bytes`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.fromPublicKey(ByteArray(16))
        }
    }

    @Test fun `fromPublicKey throws for key longer than 32 bytes`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.fromPublicKey(ByteArray(64))
        }
    }

    @Test fun `fromPublicKey throws for empty key`() {
        assertFailsWith<IllegalArgumentException> {
            AetherTag.fromPublicKey(ByteArray(0))
        }
    }

    // ── Sign-extension guard: high-byte keys must not corrupt bit packing ─────

    @Test fun `fromPublicKey handles keys with high-bit bytes without sign-extension`() {
        // All 0xFF bytes — every byte has the high bit set; without .and(0xFF) the
        // toLong() call would sign-extend and corrupt the 50-bit window.
        val key = ByteArray(32) { 0xFF.toByte() }
        val tag = AetherTag.fromPublicKey(key)
        // Just verify it produces a well-formed tag — no exception, correct format
        assertEquals(11, tag.value.length)
        assertEquals('-', tag.value[5])
    }

    @Test fun `fromPublicKey key with 0x80 in first byte produces correct format`() {
        val key = ByteArray(32) { 0 }.also { it[0] = 0x80.toByte() }
        val tag = AetherTag.fromPublicKey(key)
        assertEquals(11, tag.value.length)
        assertEquals('-', tag.value[5])
    }
}
